File: src/mobile_robot_controller/src/perfect_odom.cpp
#include "mobile_robot_controller/perfect_odom.hpp"

PerfectOdom::PerfectOdom() :
    Node("perfect_odom"), wheel_radius(0.0), wheel_separation(0.0),
    last_wheel_pos(0, 0), robot_pos(0, 0), robot_orein(0), robot_speed(0, 0), last_time(now())
{
    declare_parameter("wheel_radius", 0.1);
    declare_parameter("wheel_separation", 0.45);

    wheel_radius = get_parameter("wheel_radius").as_double();
    wheel_separation = get_parameter("wheel_separation").as_double();

    RCLCPP_INFO_STREAM(get_logger(), "wheel radius: " << wheel_radius);
    RCLCPP_INFO_STREAM(get_logger(), "wheel separation: " << wheel_separation);

    wheel_vel_to_cmd_vel_matrix << wheel_radius / 2.0, wheel_radius / 2.0, 
                                   wheel_radius / wheel_separation, - wheel_radius / wheel_separation;

    cmd_vel_to_wheel_vel_matrix = wheel_vel_to_cmd_vel_matrix.inverse();

    RCLCPP_INFO_STREAM(get_logger(), "conversion matrix: /n" << wheel_vel_to_cmd_vel_matrix);
    RCLCPP_INFO_STREAM(get_logger(), "conversion matrix inverse: /n" << cmd_vel_to_wheel_vel_matrix);

    joint_states_sub = create_subscription<JointState>(
        "joint_states",
        10,
        [this](const JointState& msg)-> void {joint_states_callback(msg);}
    );

    odom_pub = create_publisher<Odometry>(
        "odom_perfect",
        10
    );

    odom_msg.header.frame_id = "odom";
    odom_msg.child_frame_id = "base_footprint_perfect";

    transform_broadcaster = std::make_shared<tf2_ros::TransformBroadcaster>(this);

    odom_base_tf.header.frame_id = "odom";
    odom_base_tf.child_frame_id = "base_footprint_perfect";

}

void PerfectOdom::joint_states_callback(const JointState& msg)
{
    auto current_time = rclcpp::Time(msg.header.stamp);
    double current_wheel_pos_left = msg.position.at(0);
    double current_wheel_pos_right = msg.position.at(1);
    Eigen::Vector<double, 2> current_wheel_pos(current_wheel_pos_right, current_wheel_pos_left);

    double dt = (current_time - last_time).nanoseconds() / 1e9;
    Eigen::Vector<double, 2> wheel_dp = current_wheel_pos - last_wheel_pos;

    last_time = current_time;
    last_wheel_pos = current_wheel_pos;

    Eigen::Vector<double, 2> wheel_speed = wheel_dp / dt;
    robot_speed = wheel_vel_to_cmd_vel_matrix * wheel_speed;

    double v = robot_speed.coeff(0);
    double w = robot_speed.coeff(1);

    Eigen::Vector<double, 2> robot_dtf = wheel_vel_to_cmd_vel_matrix * wheel_dp;
    double ds = robot_dtf.coeff(0);
    double dtheta = robot_dtf.coeff(1);

    double theta = robot_orein = robot_orein + dtheta;
    if (robot_orein >= 2 * M_PI)
    {
        theta = robot_orein = robot_orein -  2 * M_PI;
    }
    else if (robot_orein <= -2 * M_PI)
    {
        theta = robot_orein = robot_orein +  2 * M_PI;
    }

    double dx = std::cos(theta) * ds;
    double dy = std::sin(theta) * ds;

    Eigen::Vector<double, 2> dp(dx, dy);
    robot_pos += dp;

    double x = robot_pos.coeff(0);
    double y = robot_pos.coeff(1);
    tf2::Quaternion q;
    q.setRPY(0, 0, theta);
    q.normalize();

    // RCLCPP_INFO_STREAM(get_logger(), "linear velocity: " << v << " angular velocity: " << w);
    // RCLCPP_INFO_STREAM(get_logger(), "x: " << x << " y: " << y << " theta: " << theta);

    odom_msg.header.stamp = now();
    odom_msg.pose.pose.position.x = x;
    odom_msg.pose.pose.position.y = y;
    odom_msg.pose.pose.orientation.x = q.x();
    odom_msg.pose.pose.orientation.y = q.y();
    odom_msg.pose.pose.orientation.z = q.z();
    odom_msg.pose.pose.orientation.w = q.w();
    odom_msg.twist.twist.linear.x = v;
    odom_msg.twist.twist.angular.z = w;

    odom_pub->publish(odom_msg);

    odom_base_tf.header.stamp = now();
    odom_base_tf.transform.translation.x = x;
    odom_base_tf.transform.translation.y = y;
    odom_base_tf.transform.rotation.x = q.x();
    odom_base_tf.transform.rotation.y = q.y();
    odom_base_tf.transform.rotation.z = q.z();
    odom_base_tf.transform.rotation.w = q.w();

    transform_broadcaster->sendTransform(odom_base_tf);

}

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PerfectOdom>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/mobile_robot_controller/mobile_robot_controller/perfect_odom.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from sensor_msgs.msg import JointState
from nav_msgs.msg import Odometry
from rclpy.constants import S_TO_NS
from rclpy.time import Time
from tf2_ros import TransformBroadcaster

import numpy as np
import math

from quaternion import Quaternion

class PerfectOdom(Node):
    def __init__(self):
        super().__init__(node_name="perfect_odom")

        self.declare_parameter('wheel_radius', 0.1)
        self.declare_parameter('wheel_sepaparion', 0.35)

        self._wheel_radius_ = self.get_parameter('wheel_radius').get_parameter_value().double_value
        self._wheel_separation_ = self.get_parameter('wheel_sepaparion').get_parameter_value().double_value

        self.get_logger().info(f"wheel radius: {self._wheel_radius_ :0.2f}")
        self.get_logger().info(f"wheel separation: {self._wheel_separation_ :0.2f}")

        self._wheel_vel_to_cmd_vel_matrix_ = np.array([
            [self._wheel_radius_ / 2.0, self._wheel_radius_ / 2.0],
            [self._wheel_radius_ / self._wheel_separation_, - self._wheel_radius_ / self._wheel_separation_]
        ])

        self._cmd_vel_to_wheel_vel_matrix_ = np.linalg.inv(self._wheel_vel_to_cmd_vel_matrix_)

        left_wheel_last_pos_ = 0.0
        right_wheel_last_pos_ = 0.0
        self._last_wheel_pos_ = np.array([
            [right_wheel_last_pos_],
            [left_wheel_last_pos_],
        ])
        self._robot_pos_ = np.zeros((2, 1))
        self._robot_orien_ = 0.0
        self._robot_speed_ = np.zeros((2, 1))
        self._last_time_ = self.get_clock().now()

        wheel_vel_to_cmd_vel_matrix_str = np.array2string(self._wheel_vel_to_cmd_vel_matrix_, precision=2, suppress_small=True)
        cmd_vel_to_wheel_vel_matrix_str = np.array2string(self._cmd_vel_to_wheel_vel_matrix_, precision=2, suppress_small=True)
        self.get_logger().info(f"conversion matrix: {wheel_vel_to_cmd_vel_matrix_str}")
        self.get_logger().info(f"conversion matrix inverse: {cmd_vel_to_wheel_vel_matrix_str}")


        self._joint_states_sub_ = self.create_subscription(
            msg_type=JointState,
            topic="joint_states",
            callback=self._joint_states_callback,
            qos_profile=10,
        )

        self._odom_pub_ = self.create_publisher(
            msg_type=Odometry,
            topic="odom_perfect",
            qos_profile=10,
        )

        self._odom_msg_ = Odometry()
        self._odom_msg_.header.frame_id = "odom"
        self._odom_msg_.child_frame_id = "base_footprint_perfect"

        self._transform_broadcaster_ = TransformBroadcaster(
            node=self
        )

        self._odom_base_tf_ = TransformStamped()
        self._odom_base_tf_.header.frame_id = "odom"
        self._odom_base_tf_.child_frame_id = "base_footprint_perfect"

    def _joint_states_callback(self, msg: JointState):
        current_time = Time.from_msg(msg.header.stamp)
        current_wheel_pos_left = msg.position[0]
        current_wheel_pos_right = msg.position[1]
        current_wheel_pos = np.array([
            [current_wheel_pos_right],
            [current_wheel_pos_left],
        ])

        dt = (current_time - self._last_time_).nanoseconds / S_TO_NS
        # dp_left = current_wheel_pos_left - self._left_wheel_last_pos_
        # dp_right = current_wheel_pos_right - self._right_wheel_last_pos_
        wheel_dp = current_wheel_pos - self._last_wheel_pos_

        # self._left_wheel_last_pos_ = current_wheel_pos_left
        # self._right_wheel_last_pos_ = current_wheel_pos_right
        self._last_time_ = current_time
        self._last_wheel_pos_ = current_wheel_pos

        # omega_left = dp_left / dt
        # omega_right = dp_right / dt
        # wheel_speed = np.array([
        #     [omega_right],
        #     [omega_left]
        # ])

        wheel_speed = wheel_dp / (dt + 1e-10)
        robot_speed = self._robot_speed_ = np.matmul(self._wheel_vel_to_cmd_vel_matrix_, wheel_speed)
        v, w = robot_speed[0, 0], robot_speed[1, 0]

        robot_dtf = np.matmul(self._wheel_vel_to_cmd_vel_matrix_, wheel_dp)
        ds, dtheta = robot_dtf[0, 0], robot_dtf[1, 0]

        theta = self._robot_orien_ = self._robot_orien_ + dtheta
        if self._robot_orien_ >= 2 * math.pi:
            theta = self._robot_orien_ = self._robot_orien_ - 2 * math.pi
        elif self._robot_orien_ <= -2 * math.pi:
            theta = self._robot_orien_ = self._robot_orien_ + 2 * math.pi

        dx, dy = math.cos(theta) * ds, math.sin(theta) * ds
        robot_dp = np.array([
            [dx],
            [dy],
        ])
        
        self._robot_pos_ += robot_dp

        x, y = self._robot_pos_[0, 0], self._robot_pos_[1, 0]
        q = Quaternion.from_euler(roll=0, pitch=0, yaw=theta)
        q.normalize()

        # self.get_logger().info(f"linear velocity: {v :0.3f} angular velocity: {w:0.3f}")
        # self.get_logger().info(f"x: {x :0.3f} y: {y:0.3f} theta: {theta}")

        self._odom_msg_.header.stamp = self.get_clock().now().to_msg()
        self._odom_msg_.pose.pose.position.x = x
        self._odom_msg_.pose.pose.position.y = y
        self._odom_msg_.pose.pose.orientation.x = q.x
        self._odom_msg_.pose.pose.orientation.y = q.y
        self._odom_msg_.pose.pose.orientation.z = q.z
        self._odom_msg_.pose.pose.orientation.w = q.w
        self._odom_msg_.twist.twist.linear.x = v
        self._odom_msg_.twist.twist.angular.z = w

        self._odom_pub_.publish(self._odom_msg_)

        self._odom_base_tf_.header.stamp = self.get_clock().now().to_msg()
        self._odom_base_tf_.transform.translation.x = x
        self._odom_base_tf_.transform.translation.y = y
        self._odom_base_tf_.transform.rotation.x = q.x
        self._odom_base_tf_.transform.rotation.y = q.y
        self._odom_base_tf_.transform.rotation.z = q.z
        self._odom_base_tf_.transform.rotation.w = q.w

        self._transform_broadcaster_.sendTransform(self._odom_base_tf_)



def main(args=None):
    try:
        rclpy.init(args=args)
        node = PerfectOdom()
        rclpy.spin(node=node)
        node.destroy_node()
        rclpy.shutdown()
    except KeyboardInterrupt:
        pass

if __name__ == '__main__':
    main()